File: 01. Software Engineering Foundations/02. Projects/13. Tic Tac Toe/board.rb
class Board
  def initialize
    @grid = Array.new(3) { Array.new(3, '_') }
  end

  def valid?(position)
    position.each do |coord|
      return false if coord.negative? || coord > 2
    end
    true
  end

  def empty?(position)
    x, y = position
    @grid[x][y] == '_'
  end

  def place_mark(position, mark)
    raise 'Invalid position' unless valid?(position) && empty?(position)

    x, y = position
    @grid[x][y] = mark
  end

  def print
    @grid.each do |row|
      puts row.join(' ')
    end
  end

  def win_row?(mark)
    @grid.any? do |row|
      row.all? { |cell| cell == mark }
    end
  end

  def win_col?(mark)
    @grid.transpose.any? do |col|
      col.all? { |cell| cell == mark }
    end
  end

  def win_diagonal?(mark)
    @grid[0][0] == mark && @grid[1][1] == mark && @grid[2][2] == mark ||
      @grid[0][2] == mark && @grid[1][1] == mark && @grid[2][0] == mark
  end

  def win?(mark)
    win_row?(mark) || win_col?(mark) || win_diagonal?(mark)
  end

  def empty_positions?
    @grid.flatten.any? { |cell| cell == '_' }
  end
end

File: 05. JavaScript/02. Projects/05. Tic-Tac-Toe JS Edition/board.js
class Board {
  constructor() {
    this.grid = this.makeGrid();
  }

  makeGrid() {
    const grid = [];
    for (let i = 0; i < 3; i++) {
      grid.push([]);
      for (let j = 0; j < 3; j++) {
        grid[i].push(null);
      }
    }
    return grid;
  }

  print() {
    this.grid.forEach(row => {
      console.log(row.join(" | "));
    });
  }

  won() {
    const winningCombos = [
      [0, 1, 2],
      [3, 4, 5],
      [6, 7, 8],
      [0, 3, 6],
      [1, 4, 7],
      [2, 5, 8],
      [0, 4, 8],
      [2, 4, 6]
    ];
    for (let i = 0; i < winningCombos.length; i++) {
      const [a, b, c] = winningCombos[i];
      if (
        this.grid[a][0] &&
        this.grid[a][0] === this.grid[b][0] &&
        this.grid[a][0] === this.grid[c][0]
      ) {
        return this.grid[a][0];
      }
    }
    return false;
  }

  winner() {
    this.won();
  }

  empty(pos) {
    return this.grid[pos[0]][pos[1]] === null;
  }

  placeMark(pos, mark) {
    if (!this.empty(pos)) {
      return false;
    }

    this.grid[pos[0]][pos[1]] = mark;
  }
}

module.exports = Board;
